{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Note: include in the tests folder a .env file that contains the API keys for the services used in the tests    \n",
    "import os\n",
    "import pandas as pd  # type: ignore\n",
    "import palimpzest as pz\n",
    "\n",
    "if not os.environ.get('OPENAI_API_KEY'):\n",
    "    with open('.env') as f:\n",
    "        for line in f:\n",
    "            key, value = line.strip().split('=')\n",
    "            os.environ[key] = value"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Columns definition\n",
    "In the following cell we target schema of the case study we are interested in. Note how we are not specifying target attributes with fine grained metadata but rather with natural language, possibly ambiguous, specifications."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Case Data columns:\n",
      "{'name': 'case_submitter_id', 'type': <class 'str'>, 'desc': 'The ID of the case'}\n",
      "{'name': 'age_at_diagnosis', 'type': int | float, 'desc': 'The age of the patient at the time of diagnosis'}\n",
      "{'name': 'race', 'type': <class 'str'>, 'desc': 'An arbitrary classification of a taxonomic group that is a division of a species.'}\n",
      "{'name': 'ethnicity', 'type': <class 'str'>, 'desc': 'Whether an individual describes themselves as Hispanic or Latino or not.'}\n",
      "{'name': 'gender', 'type': <class 'str'>, 'desc': 'Text designations that identify gender.'}\n"
     ]
    }
   ],
   "source": [
    "case_data_cols = [\n",
    "    {\"name\": \"case_submitter_id\", \"type\": str, \"desc\": \"The ID of the case\"},\n",
    "    {\"name\": \"age_at_diagnosis\", \"type\": int | float, \"desc\": \"The age of the patient at the time of diagnosis\"},\n",
    "    {\"name\": \"race\", \"type\": str, \"desc\": \"An arbitrary classification of a taxonomic group that is a division of a species.\"},\n",
    "    {\"name\": \"ethnicity\", \"type\": str, \"desc\": \"Whether an individual describes themselves as Hispanic or Latino or not.\"},\n",
    "    {\"name\": \"gender\", \"type\": str, \"desc\": \"Text designations that identify gender.\"},\n",
    "    {\"name\": \"vital_status\", \"type\": str, \"desc\": \"The vital status of the patient\"},\n",
    "    {\"name\": \"ajcc_pathologic_t\", \"type\": str, \"desc\": \"Code of pathological T (primary tumor) to define the size or contiguous extension of the primary tumor (T), using staging criteria from the American Joint Committee on Cancer (AJCC).\"},\n",
    "    {\"name\": \"ajcc_pathologic_n\", \"type\": str, \"desc\": \"The codes that represent the stage of cancer based on the nodes present (N stage) according to criteria based on multiple editions of the AJCC's Cancer Staging Manual.\"},\n",
    "    {\"name\": \"ajcc_pathologic_stage\", \"type\": str, \"desc\": \"The extent of a cancer, especially whether the disease has spread from the original site to other parts of the body based on AJCC staging criteria.\"},\n",
    "    {\"name\": \"tumor_grade\", \"type\": int | float, \"desc\": \"Numeric value to express the degree of abnormality of cancer cells, a measure of differentiation and aggressiveness.\"},\n",
    "    {\"name\": \"tumor_focality\", \"type\": str, \"desc\": \"The text term used to describe whether the patient's disease originated in a single location or multiple locations.\"},\n",
    "    {\"name\": \"tumor_largest_dimension_diameter\", \"type\": int | float, \"desc\": \"The tumor largest dimension diameter.\"},\n",
    "    {\"name\": \"primary_diagnosis\", \"type\": str, \"desc\": \"Text term used to describe the patient's histologic diagnosis, as described by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).\"},\n",
    "    {\"name\": \"morphology\", \"type\": str, \"desc\": \"The Morphological code of the tumor, as described by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).\"},\n",
    "    {\"name\": \"tissue_or_organ_of_origin\", \"type\": str, \"desc\": \"The text term used to describe the anatomic site of origin, of the patient's malignant disease, as described by the World Health Organization's (WHO) International Classification of Diseases for Oncology (ICD-O).\"},\n",
    "    {\"name\": \"study\", \"type\": str, \"desc\": \"The last name of the author of the study, from the table name\"}\n",
    "]\n",
    "\n",
    "file_cols = [\n",
    "    {\"name\": \"filename\", \"type\": str, \"desc\": \"The name of the file\"},\n",
    "    {\"name\": \"contents\", \"type\": bytes, \"desc\": \"The contents of the file\"}\n",
    "]\n",
    "\n",
    "xls_cols = file_cols + [\n",
    "    {\"name\": \"number_sheets\", \"type\": int, \"desc\": \"The number of sheets in the Excel file\"},\n",
    "    {\"name\": \"sheet_names\", \"type\": list[str], \"desc\": \"The names of the sheets in the Excel file\"},\n",
    "]\n",
    "\n",
    "table_cols = [\n",
    "    {\"name\": \"rows\", \"type\": list[str], \"desc\": \"The rows of the table\"},\n",
    "    {\"name\": \"header\", \"type\": list[str], \"desc\": \"The header of the table\"},\n",
    "    {\"name\": \"name\", \"type\": str, \"desc\": \"The name of the table\"},\n",
    "    {\"name\": \"filename\", \"type\": str, \"desc\": \"The name of the file the table was extracted from\"}\n",
    "]\n",
    "\n",
    "\n",
    "print(\"Case Data columns:\")\n",
    "for col in case_data_cols[:5]:\n",
    "    print(col)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "from palimpzest.core.elements.records import DataRecord\n",
    "\n",
    "def print_result(records):\n",
    "    df = DataRecord.to_df(records)\n",
    "    cols = records[0].schema.field_names()\n",
    "    display(df[cols])\n",
    "\n",
    "def print_stats(stats):\n",
    "    print(\"Chosen plan\")\n",
    "    print(result.executed_plans[0])\n",
    "    print(\"Total execution time:\", stats.total_execution_time,\"seconds\")\n",
    "    print(\"Total execution cost:\", stats.total_execution_cost, \"USD\")\n",
    "    # print(\"Total time optimizing pipeline:\", stats.total_optimization_time, \"seconds\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Printing base dataset\n",
    "In the following cell we print the base dataset we are interested in. Note how we are not specifying target attributes with fine grained metadata but rather with natural language, possibly ambiguous, specifications."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "6c697da664c141eab26e8f21af475d69",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(IntProgress(value=0, bar_style='info', description='Processing:', max=11), HTML(value='<pre>Iniâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filename</th>\n",
       "      <th>html</th>\n",
       "      <th>text</th>\n",
       "      <th>timestamp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.724867e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>clark.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Integrated Proteogenomic Characterization of C...</td>\n",
       "      <td>1.724867e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>dou.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic Characterization of Endometrial ...</td>\n",
       "      <td>1.724867e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>gilette.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic Characterization Reveals Therape...</td>\n",
       "      <td>1.724867e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>huang.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic insights into the biology and tr...</td>\n",
       "      <td>1.724867e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>krug.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic Landscape of Breast Cancer Tumor...</td>\n",
       "      <td>1.724867e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>li.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic data and resources for pan-cance...</td>\n",
       "      <td>1.724867e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>mcdermott.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic Characterization of Ovarian HGSC...</td>\n",
       "      <td>1.724867e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>satpathy.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>A proteogenomic portrait of lung squamous cell...</td>\n",
       "      <td>1.724867e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>vasaikar.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic Analysis of Human Colon Cancer R...</td>\n",
       "      <td>1.724867e+09</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>wang.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic and metabolomic characterization...</td>\n",
       "      <td>1.724867e+09</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         filename                                               html  \\\n",
       "0         cao.htm  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "1       clark.htm  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "2         dou.htm  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "3     gilette.htm  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "4       huang.htm  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "5        krug.htm  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "6          li.htm  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "7   mcdermott.htm  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "8    satpathy.htm  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "9    vasaikar.htm  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "10       wang.htm  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "\n",
       "                                                 text     timestamp  \n",
       "0   Proteogenomic characterization of pancreatic d...  1.724867e+09  \n",
       "1   Integrated Proteogenomic Characterization of C...  1.724867e+09  \n",
       "2   Proteogenomic Characterization of Endometrial ...  1.724867e+09  \n",
       "3   Proteogenomic Characterization Reveals Therape...  1.724867e+09  \n",
       "4   Proteogenomic insights into the biology and tr...  1.724867e+09  \n",
       "5   Proteogenomic Landscape of Breast Cancer Tumor...  1.724867e+09  \n",
       "6   Proteogenomic data and resources for pan-cance...  1.724867e+09  \n",
       "7   Proteogenomic Characterization of Ovarian HGSC...  1.724867e+09  \n",
       "8   A proteogenomic portrait of lung squamous cell...  1.724867e+09  \n",
       "9   Proteogenomic Analysis of Human Colon Cancer R...  1.724867e+09  \n",
       "10  Proteogenomic and metabolomic characterization...  1.724867e+09  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "papers_html = pz.Dataset(\"../testdata/biofabric-html\")\n",
    "output = papers_html\n",
    "\n",
    "policy = pz.MinCost()\n",
    "config = pz.QueryProcessorConfig(\n",
    "    policy=policy,\n",
    "    nocache=True,\n",
    "    processing_strategy=\"no_sentinel\",\n",
    ")\n",
    "result = output.run(config)\n",
    "\n",
    "print_result(result.data_records)\n",
    "# print_stats(result.execution_stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Downloading relevant tables from the papers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "5b6177cf853c4566bb1c5c5737f6183f",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(IntProgress(value=0, bar_style='info', description='Processing:', max=3), HTML(value='<pre>Initâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contents</th>\n",
       "      <th>filename</th>\n",
       "      <th>html</th>\n",
       "      <th>text</th>\n",
       "      <th>timestamp</th>\n",
       "      <th>url</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b''</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/d1c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/0f1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/40d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/01c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/9be...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/a1c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/39e...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>b''</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>[]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/d1c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/0f1...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/40d...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/01c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/9be...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/a1c...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>b'&lt;!DOCTYPE html&gt;&lt;html lang=\"en-US\"&gt;&lt;head&gt;&lt;tit...</td>\n",
       "      <td>cao.htm</td>\n",
       "      <td>&lt;!DOCTYPE html&gt; &lt;html lang=\"en\" class=\"pb-page...</td>\n",
       "      <td>Proteogenomic characterization of pancreatic d...</td>\n",
       "      <td>1.742491e+09</td>\n",
       "      <td>/cms/10.1016/j.cell.2021.08.023/attachment/39e...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                             contents filename  \\\n",
       "0                                                 b''  cao.htm   \n",
       "1   b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "2   b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "3   b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "4   b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "5   b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "6   b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "7   b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "8                                                 b''  cao.htm   \n",
       "9   b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "10  b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "11  b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "12  b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "13  b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "14  b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "15  b'<!DOCTYPE html><html lang=\"en-US\"><head><tit...  cao.htm   \n",
       "\n",
       "                                                 html  \\\n",
       "0   <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "1   <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "2   <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "3   <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "4   <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "5   <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "6   <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "7   <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "8   <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "9   <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "10  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "11  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "12  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "13  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "14  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "15  <!DOCTYPE html> <html lang=\"en\" class=\"pb-page...   \n",
       "\n",
       "                                                 text     timestamp  \\\n",
       "0   Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "1   Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "2   Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "3   Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "4   Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "5   Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "6   Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "7   Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "8   Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "9   Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "10  Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "11  Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "12  Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "13  Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "14  Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "15  Proteogenomic characterization of pancreatic d...  1.742491e+09   \n",
       "\n",
       "                                                  url  \n",
       "0                                                  []  \n",
       "1   /cms/10.1016/j.cell.2021.08.023/attachment/d1c...  \n",
       "2   /cms/10.1016/j.cell.2021.08.023/attachment/0f1...  \n",
       "3   /cms/10.1016/j.cell.2021.08.023/attachment/40d...  \n",
       "4   /cms/10.1016/j.cell.2021.08.023/attachment/01c...  \n",
       "5   /cms/10.1016/j.cell.2021.08.023/attachment/9be...  \n",
       "6   /cms/10.1016/j.cell.2021.08.023/attachment/a1c...  \n",
       "7   /cms/10.1016/j.cell.2021.08.023/attachment/39e...  \n",
       "8                                                  []  \n",
       "9   /cms/10.1016/j.cell.2021.08.023/attachment/d1c...  \n",
       "10  /cms/10.1016/j.cell.2021.08.023/attachment/0f1...  \n",
       "11  /cms/10.1016/j.cell.2021.08.023/attachment/40d...  \n",
       "12  /cms/10.1016/j.cell.2021.08.023/attachment/01c...  \n",
       "13  /cms/10.1016/j.cell.2021.08.023/attachment/9be...  \n",
       "14  /cms/10.1016/j.cell.2021.08.023/attachment/a1c...  \n",
       "15  /cms/10.1016/j.cell.2021.08.023/attachment/39e...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "from palimpzest.utils import udfs\n",
    "\n",
    "papers_html = pz.Dataset(\"../testdata/biofabric-html-tiny\")\n",
    "table_urls = papers_html.sem_add_columns(\n",
    "    [{\"name\": \"url\", \"type\": str, \"desc\": \"The URLs of the XLS tables from the page\"}],\n",
    "    cardinality=pz.Cardinality.ONE_TO_MANY)\n",
    "files = table_urls.download()\n",
    "output = files\n",
    "\n",
    "policy = pz.MinCost()\n",
    "config = pz.QueryProcessorConfig(policy=policy, nocache=True)\n",
    "result = output.run(config)\n",
    "\n",
    "print_result(result.data_records)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filtering stage\n",
    "In the following cell we define the first part of the workload, that comprises a filtering stage responsible for selecting the tables from all the spreadsheets that contain relevant biometric information about the patient.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "33b0b6c927e144d0bd071265e96da8d2",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(IntProgress(value=0, bar_style='info', description='Processing:', max=9), HTML(value='<pre>Initâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contents</th>\n",
       "      <th>filename</th>\n",
       "      <th>header</th>\n",
       "      <th>name</th>\n",
       "      <th>number_sheets</th>\n",
       "      <th>rows</th>\n",
       "      <th>sheet_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>dou_mmc1.xlsx_UCEC_CPTAC3_meta_table_V2.1</td>\n",
       "      <td>1</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>[Column name, Column definition, Unnamed: 2, U...</td>\n",
       "      <td>vasaikar_mmc1.xlsx_A-Annotation</td>\n",
       "      <td>3</td>\n",
       "      <td>[[SampleID, ID of the prospective participant,...</td>\n",
       "      <td>[Description, A-Annotation, B-ClinicalData]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>[SampleID, Clinical, Mutation, CNV, miRNA-seq,...</td>\n",
       "      <td>vasaikar_mmc1.xlsx_B-ClinicalData</td>\n",
       "      <td>3</td>\n",
       "      <td>[[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...</td>\n",
       "      <td>[Description, A-Annotation, B-ClinicalData]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>[case_id, age, gender, height, weight, bmi, co...</td>\n",
       "      <td>wang_mmc2.xlsx_clinical_data</td>\n",
       "      <td>5</td>\n",
       "      <td>[[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>[case, sample_type, multiomic, nmf_consensus, ...</td>\n",
       "      <td>wang_mmc2.xlsx_additional_annotations</td>\n",
       "      <td>5</td>\n",
       "      <td>[[C3L-00104, tumor, nmf1, nmf1, 0.743, Proneur...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>[aliquot_id, sample_id, original_aliquot_id, b...</td>\n",
       "      <td>wang_mmc2.xlsx_cbttc_sample_info</td>\n",
       "      <td>5</td>\n",
       "      <td>[[7316-288-TISS [894482], 7316-288, 7316-288-T...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            contents            filename  \\\n",
       "0  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...       dou_mmc1.xlsx   \n",
       "1  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  vasaikar_mmc1.xlsx   \n",
       "2  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  vasaikar_mmc1.xlsx   \n",
       "3  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...      wang_mmc2.xlsx   \n",
       "4  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...      wang_mmc2.xlsx   \n",
       "5  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...      wang_mmc2.xlsx   \n",
       "\n",
       "                                              header  \\\n",
       "0  [idx, Proteomics_Participant_ID, Case_excluded...   \n",
       "1  [Column name, Column definition, Unnamed: 2, U...   \n",
       "2  [SampleID, Clinical, Mutation, CNV, miRNA-seq,...   \n",
       "3  [case_id, age, gender, height, weight, bmi, co...   \n",
       "4  [case, sample_type, multiomic, nmf_consensus, ...   \n",
       "5  [aliquot_id, sample_id, original_aliquot_id, b...   \n",
       "\n",
       "                                        name  number_sheets  \\\n",
       "0  dou_mmc1.xlsx_UCEC_CPTAC3_meta_table_V2.1              1   \n",
       "1            vasaikar_mmc1.xlsx_A-Annotation              3   \n",
       "2          vasaikar_mmc1.xlsx_B-ClinicalData              3   \n",
       "3               wang_mmc2.xlsx_clinical_data              5   \n",
       "4      wang_mmc2.xlsx_additional_annotations              5   \n",
       "5           wang_mmc2.xlsx_cbttc_sample_info              5   \n",
       "\n",
       "                                                rows  \\\n",
       "0  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "1  [[SampleID, ID of the prospective participant,...   \n",
       "2  [[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...   \n",
       "3  [[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...   \n",
       "4  [[C3L-00104, tumor, nmf1, nmf1, 0.743, Proneur...   \n",
       "5  [[7316-288-TISS [894482], 7316-288, 7316-288-T...   \n",
       "\n",
       "                                         sheet_names  \n",
       "0                      [UCEC_CPTAC3_meta_table_V2.1]  \n",
       "1        [Description, A-Annotation, B-ClinicalData]  \n",
       "2        [Description, A-Annotation, B-ClinicalData]  \n",
       "3  [README, clinical_data, additional_annotations...  \n",
       "4  [README, clinical_data, additional_annotations...  \n",
       "5  [README, clinical_data, additional_annotations...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chosen plan\n",
      " 0. MarshalAndScanDataOp -> XLSFile \n",
      "\n",
      " 1. XLSFile -> NonLLMConvert -> Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']]\n",
      "    (contents, filename, number_she) -> (contents, filename, header, na)\n",
      "    UDF: <function xls_to_tables at 0x7fcdc4214720>\n",
      "\n",
      " 2. Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']] -> LLMFilter -> Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']]\n",
      "    (contents, filename, header, na) -> (contents, filename, header, na)\n",
      "    Model: Model.GPT_4o_MINI\n",
      "    Filter: The table contains biometric information about the patient\n",
      "\n",
      "\n",
      "Total execution time: 15.247003078460693 seconds\n",
      "Total execution cost: 0.0029549999999999993 USD\n"
     ]
    }
   ],
   "source": [
    "from palimpzest.utils import udfs\n",
    "xls = pz.Dataset(\"../testdata/biofabric-tiny\")\n",
    "patient_tables = xls.add_columns(udf=udfs.xls_to_tables, cols=table_cols, cardinality=pz.Cardinality.ONE_TO_MANY)\n",
    "patient_tables = patient_tables.sem_filter(\"The table contains biometric information about the patient\")\n",
    "\n",
    "output = patient_tables\n",
    "\n",
    "policy = pz.MinCost()\n",
    "config = pz.QueryProcessorConfig(policy=policy,nocache=True)\n",
    "result = patient_tables.run(config)\n",
    "\n",
    "print_result(result.data_records)\n",
    "print_stats(result.execution_stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "27620b9d819f45609a1db991e6715edc",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(IntProgress(value=0, bar_style='info', description='Processing:', max=9), HTML(value='<pre>Initâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contents</th>\n",
       "      <th>filename</th>\n",
       "      <th>header</th>\n",
       "      <th>name</th>\n",
       "      <th>number_sheets</th>\n",
       "      <th>rows</th>\n",
       "      <th>sheet_names</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>dou_mmc1.xlsx_UCEC_CPTAC3_meta_table_V2.1</td>\n",
       "      <td>1</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>[SampleID, Clinical, Mutation, CNV, miRNA-seq,...</td>\n",
       "      <td>vasaikar_mmc1.xlsx_B-ClinicalData</td>\n",
       "      <td>3</td>\n",
       "      <td>[[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...</td>\n",
       "      <td>[Description, A-Annotation, B-ClinicalData]</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>[README for Table S1, Unnamed: 1, Unnamed: 2]</td>\n",
       "      <td>wang_mmc2.xlsx_README</td>\n",
       "      <td>5</td>\n",
       "      <td>[[nan, nan, nan], [Tab1, clinical_data, Clinic...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>[case_id, age, gender, height, weight, bmi, co...</td>\n",
       "      <td>wang_mmc2.xlsx_clinical_data</td>\n",
       "      <td>5</td>\n",
       "      <td>[[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                            contents            filename  \\\n",
       "0  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...       dou_mmc1.xlsx   \n",
       "1  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  vasaikar_mmc1.xlsx   \n",
       "2  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...      wang_mmc2.xlsx   \n",
       "3  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...      wang_mmc2.xlsx   \n",
       "\n",
       "                                              header  \\\n",
       "0  [idx, Proteomics_Participant_ID, Case_excluded...   \n",
       "1  [SampleID, Clinical, Mutation, CNV, miRNA-seq,...   \n",
       "2      [README for Table S1, Unnamed: 1, Unnamed: 2]   \n",
       "3  [case_id, age, gender, height, weight, bmi, co...   \n",
       "\n",
       "                                        name  number_sheets  \\\n",
       "0  dou_mmc1.xlsx_UCEC_CPTAC3_meta_table_V2.1              1   \n",
       "1          vasaikar_mmc1.xlsx_B-ClinicalData              3   \n",
       "2                      wang_mmc2.xlsx_README              5   \n",
       "3               wang_mmc2.xlsx_clinical_data              5   \n",
       "\n",
       "                                                rows  \\\n",
       "0  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "1  [[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...   \n",
       "2  [[nan, nan, nan], [Tab1, clinical_data, Clinic...   \n",
       "3  [[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...   \n",
       "\n",
       "                                         sheet_names  \n",
       "0                      [UCEC_CPTAC3_meta_table_V2.1]  \n",
       "1        [Description, A-Annotation, B-ClinicalData]  \n",
       "2  [README, clinical_data, additional_annotations...  \n",
       "3  [README, clinical_data, additional_annotations...  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chosen plan\n",
      " 0. MarshalAndScanDataOp -> XLSFile \n",
      "\n",
      " 1. XLSFile -> NonLLMConvert -> Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']]\n",
      "    (contents, filename, number_she) -> (contents, filename, header, na)\n",
      "    UDF: <function xls_to_tables at 0x7fcdc4214720>\n",
      "\n",
      " 2. Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']] -> LLMFilter -> Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']]\n",
      "    (contents, filename, header, na) -> (contents, filename, header, na)\n",
      "    Model: Model.GPT_4o\n",
      "    Filter: The table contains biometric information about the patient\n",
      "\n",
      "\n",
      "Total execution time: 20.54369831085205 seconds\n",
      "Total execution cost: 0.05419000000000001 USD\n"
     ]
    }
   ],
   "source": [
    "policy = pz.MaxQuality()\n",
    "\n",
    "config = pz.QueryProcessorConfig(policy=policy, nocache=True)\n",
    "result = patient_tables.run(config)\n",
    "print_result(result.data_records)\n",
    "print_stats(result.execution_stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Matching stage\n",
    "Once filtered, we can define the second part of the workload, that matches the patient records in the different tables and merging the information into a single table."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Minimum cost: GPT 3.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "09efb5b903c84faeb9a6a4aad3470a55",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(IntProgress(value=0, bar_style='info', description='Processing:', max=12), HTML(value='<pre>Iniâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>ajcc_pathologic_n</th>\n",
       "      <th>ajcc_pathologic_stage</th>\n",
       "      <th>ajcc_pathologic_t</th>\n",
       "      <th>case_submitter_id</th>\n",
       "      <th>contents</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>filename</th>\n",
       "      <th>gender</th>\n",
       "      <th>header</th>\n",
       "      <th>...</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>race</th>\n",
       "      <th>rows</th>\n",
       "      <th>sheet_names</th>\n",
       "      <th>study</th>\n",
       "      <th>tissue_or_organ_of_origin</th>\n",
       "      <th>tumor_focality</th>\n",
       "      <th>tumor_grade</th>\n",
       "      <th>tumor_largest_dimension_diameter</th>\n",
       "      <th>vital_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>38.88</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00006</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>...</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>White</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "      <td>dou</td>\n",
       "      <td>Anterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>2.9</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39.76</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00008</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>...</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>White</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "      <td>dou</td>\n",
       "      <td>Posterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>3.5</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>51.19</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00032</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>...</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>White</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "      <td>dou</td>\n",
       "      <td>Anterior and Posterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>4.5</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C3L-00084</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>No</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>...</td>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "      <td>dou</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32.69</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00090</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>...</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>White</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "      <td>dou</td>\n",
       "      <td>Anterior and Posterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>3.5</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>None</td>\n",
       "      <td>[Sheet, Description]</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[A-Annotation, Column annotations], [B-Clinic...</td>\n",
       "      <td>[Description, B-ClinicalData]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>729.00</td>\n",
       "      <td>N2b</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>T4a</td>\n",
       "      <td>01CO001</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>Male</td>\n",
       "      <td>[SampleID, Clinical, Mutation, CNV, miRNA-seq,...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...</td>\n",
       "      <td>[Description, B-ClinicalData]</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Living</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>838.00</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>T3</td>\n",
       "      <td>01CO005</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[SampleID, Clinical, Mutation, CNV, miRNA-seq,...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...</td>\n",
       "      <td>[Description, B-ClinicalData]</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>904.00</td>\n",
       "      <td>N2b</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>T4a</td>\n",
       "      <td>01CO006</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[SampleID, Clinical, Mutation, CNV, miRNA-seq,...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...</td>\n",
       "      <td>[Description, B-ClinicalData]</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Ascending Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Living</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>652.00</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>T3</td>\n",
       "      <td>01CO008</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[SampleID, Clinical, Mutation, CNV, miRNA-seq,...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...</td>\n",
       "      <td>[Description, B-ClinicalData]</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Descending Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Living</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>58.00</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>C3L-00104</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>Male</td>\n",
       "      <td>[case_id, age, gender, height, weight, bmi, co...</td>\n",
       "      <td>...</td>\n",
       "      <td>GBM</td>\n",
       "      <td>White</td>\n",
       "      <td>[[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...</td>\n",
       "      <td>[clinical_data]</td>\n",
       "      <td>Wang</td>\n",
       "      <td>Frontal Lobe</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>nan</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>11 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    age_at_diagnosis ajcc_pathologic_n ajcc_pathologic_stage  \\\n",
       "0              38.88               pN0               Stage I   \n",
       "1              39.76               pNX              Stage IV   \n",
       "2              51.19               pN0               Stage I   \n",
       "3                NaN              None                  None   \n",
       "4              32.69               pNX               Stage I   \n",
       "5                NaN              None                  None   \n",
       "6             729.00               N2b             Stage III   \n",
       "7             838.00                N0              Stage II   \n",
       "8             904.00               N2b             Stage III   \n",
       "9             652.00                N0              Stage II   \n",
       "10             58.00               nan                   nan   \n",
       "\n",
       "   ajcc_pathologic_t case_submitter_id  \\\n",
       "0     pT1a (FIGO IA)         C3L-00006   \n",
       "1     pT1a (FIGO IA)         C3L-00008   \n",
       "2     pT1a (FIGO IA)         C3L-00032   \n",
       "3               None         C3L-00084   \n",
       "4     pT1a (FIGO IA)         C3L-00090   \n",
       "5               None              None   \n",
       "6                T4a           01CO001   \n",
       "7                 T3           01CO005   \n",
       "8                T4a           01CO006   \n",
       "9                 T3           01CO008   \n",
       "10               nan         C3L-00104   \n",
       "\n",
       "                                             contents               ethnicity  \\\n",
       "0   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  Not-Hispanic or Latino   \n",
       "1   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  Not-Hispanic or Latino   \n",
       "2   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  Not-Hispanic or Latino   \n",
       "3   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                      No   \n",
       "4   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  Not-Hispanic or Latino   \n",
       "5   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "6   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "7   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "8   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "9   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "10  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  Not-Hispanic or Latino   \n",
       "\n",
       "              filename  gender  \\\n",
       "0        dou_mmc1.xlsx  Female   \n",
       "1        dou_mmc1.xlsx  Female   \n",
       "2        dou_mmc1.xlsx  Female   \n",
       "3        dou_mmc1.xlsx  Female   \n",
       "4        dou_mmc1.xlsx  Female   \n",
       "5   vasaikar_mmc1.xlsx    None   \n",
       "6   vasaikar_mmc1.xlsx    Male   \n",
       "7   vasaikar_mmc1.xlsx  Female   \n",
       "8   vasaikar_mmc1.xlsx  Female   \n",
       "9   vasaikar_mmc1.xlsx  Female   \n",
       "10      wang_mmc2.xlsx    Male   \n",
       "\n",
       "                                               header  ... primary_diagnosis  \\\n",
       "0   [idx, Proteomics_Participant_ID, Case_excluded...  ...      Endometrioid   \n",
       "1   [idx, Proteomics_Participant_ID, Case_excluded...  ...      Endometrioid   \n",
       "2   [idx, Proteomics_Participant_ID, Case_excluded...  ...      Endometrioid   \n",
       "3   [idx, Proteomics_Participant_ID, Case_excluded...  ...    Carcinosarcoma   \n",
       "4   [idx, Proteomics_Participant_ID, Case_excluded...  ...      Endometrioid   \n",
       "5                                [Sheet, Description]  ...              None   \n",
       "6   [SampleID, Clinical, Mutation, CNV, miRNA-seq,...  ...              None   \n",
       "7   [SampleID, Clinical, Mutation, CNV, miRNA-seq,...  ...              None   \n",
       "8   [SampleID, Clinical, Mutation, CNV, miRNA-seq,...  ...              None   \n",
       "9   [SampleID, Clinical, Mutation, CNV, miRNA-seq,...  ...              None   \n",
       "10  [case_id, age, gender, height, weight, bmi, co...  ...               GBM   \n",
       "\n",
       "     race                                               rows  \\\n",
       "0   White  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "1   White  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "2   White  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "3    None  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "4   White  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "5    None  [[A-Annotation, Column annotations], [B-Clinic...   \n",
       "6    None  [[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...   \n",
       "7    None  [[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...   \n",
       "8    None  [[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...   \n",
       "9    None  [[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...   \n",
       "10  White  [[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...   \n",
       "\n",
       "                      sheet_names     study  \\\n",
       "0   [UCEC_CPTAC3_meta_table_V2.1]       dou   \n",
       "1   [UCEC_CPTAC3_meta_table_V2.1]       dou   \n",
       "2   [UCEC_CPTAC3_meta_table_V2.1]       dou   \n",
       "3   [UCEC_CPTAC3_meta_table_V2.1]       dou   \n",
       "4   [UCEC_CPTAC3_meta_table_V2.1]       dou   \n",
       "5   [Description, B-ClinicalData]      None   \n",
       "6   [Description, B-ClinicalData]  Vasaikar   \n",
       "7   [Description, B-ClinicalData]  Vasaikar   \n",
       "8   [Description, B-ClinicalData]  Vasaikar   \n",
       "9   [Description, B-ClinicalData]  Vasaikar   \n",
       "10                [clinical_data]      Wang   \n",
       "\n",
       "             tissue_or_organ_of_origin tumor_focality   tumor_grade  \\\n",
       "0                 Anterior endometrium       Unifocal  FIGO grade 1   \n",
       "1                Posterior endometrium       Unifocal  FIGO grade 1   \n",
       "2   Anterior and Posterior endometrium       Unifocal  FIGO grade 2   \n",
       "3                                 None           None          None   \n",
       "4   Anterior and Posterior endometrium       Unifocal  FIGO grade 2   \n",
       "5                                 None           None          None   \n",
       "6                        Sigmoid Colon           None          None   \n",
       "7                        Sigmoid Colon           None          None   \n",
       "8                      Ascending Colon           None          None   \n",
       "9                     Descending Colon           None          None   \n",
       "10                        Frontal Lobe            nan           nan   \n",
       "\n",
       "   tumor_largest_dimension_diameter vital_status  \n",
       "0                               2.9           No  \n",
       "1                               3.5           No  \n",
       "2                               4.5           No  \n",
       "3                              None          Yes  \n",
       "4                               3.5           No  \n",
       "5                              None         None  \n",
       "6                              None       Living  \n",
       "7                              None     Deceased  \n",
       "8                              None       Living  \n",
       "9                              None       Living  \n",
       "10                              nan     Deceased  \n",
       "\n",
       "[11 rows x 23 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chosen plan\n",
      " 0. MarshalAndScanDataOp -> XLSFile \n",
      "\n",
      " 1. XLSFile -> NonLLMConvert -> Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']]\n",
      "    (contents, filename, number_she) -> (contents, filename, header, na)\n",
      "    UDF: <function xls_to_tables at 0x7faeec21bba0>\n",
      "\n",
      " 2. Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']] -> LLMFilter -> Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']]\n",
      "    (contents, filename, header, na) -> (contents, filename, header, na)\n",
      "    Model: Model.GPT_4o_MINI\n",
      "    Filter: The table contains biometric information about the patient\n",
      "\n",
      "\n",
      "Total execution time: 50.044567823410034 seconds\n",
      "Total execution cost: 0.0061902 USD\n"
     ]
    }
   ],
   "source": [
    "from palimpzest.utils import udfs   \n",
    "input_dataset = pz.Dataset(\"../testdata/biofabric-tiny-filtered\")\n",
    "patient_tables = input_dataset.add_columns(udf=udfs.xls_to_tables, cols=table_cols, cardinality=pz.Cardinality.ONE_TO_MANY)\n",
    "patient_tables = patient_tables.sem_filter(\"The table contains biometric information about the patient\")\n",
    "case_data = patient_tables.sem_add_columns(case_data_cols, cardinality=pz.Cardinality.ONE_TO_MANY)\n",
    "\n",
    "policy = pz.MinCost()\n",
    "config = pz.QueryProcessorConfig(\n",
    "    policy=policy,\n",
    "    nocache=True,\n",
    "    allow_code_synth=False,\n",
    "    processing_strategy=\"no_sentinel\",\n",
    ")\n",
    "data_record_collection = case_data.run(config) \n",
    "\n",
    "print_result(data_record_collection.data_records)\n",
    "print_stats(data_record_collection.execution_stats)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## End to end"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Minimum cost : GPT 3.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.jupyter.widget-view+json": {
       "model_id": "273458e7c4224a6d83b2ffc2622fc350",
       "version_major": 2,
       "version_minor": 0
      },
      "text/plain": [
       "VBox(children=(IntProgress(value=0, bar_style='info', description='Processing:', max=12), HTML(value='<pre>Iniâ€¦"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Error parsing answers: Extra data: line 1 column 266 (char 265)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>ajcc_pathologic_n</th>\n",
       "      <th>ajcc_pathologic_stage</th>\n",
       "      <th>ajcc_pathologic_t</th>\n",
       "      <th>case_submitter_id</th>\n",
       "      <th>contents</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>filename</th>\n",
       "      <th>gender</th>\n",
       "      <th>header</th>\n",
       "      <th>...</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>race</th>\n",
       "      <th>rows</th>\n",
       "      <th>sheet_names</th>\n",
       "      <th>study</th>\n",
       "      <th>tissue_or_organ_of_origin</th>\n",
       "      <th>tumor_focality</th>\n",
       "      <th>tumor_grade</th>\n",
       "      <th>tumor_largest_dimension_diameter</th>\n",
       "      <th>vital_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>38.88</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00006</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>...</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>White</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "      <td>dou_mmc1</td>\n",
       "      <td>Anterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>2.9</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>39.76</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00008</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>...</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>White</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "      <td>dou_mmc1</td>\n",
       "      <td>Posterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>3.5</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>51.19</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00032</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>...</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>White</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "      <td>dou_mmc1</td>\n",
       "      <td>Anterior and Posterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>4.5</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C3L-00084</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>No</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>None</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>...</td>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "      <td>dou_mmc1</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>32.69</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00090</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>dou_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[idx, Proteomics_Participant_ID, Case_excluded...</td>\n",
       "      <td>...</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>White</td>\n",
       "      <td>[[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...</td>\n",
       "      <td>[UCEC_CPTAC3_meta_table_V2.1]</td>\n",
       "      <td>dou_mmc1</td>\n",
       "      <td>Anterior and Posterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>3.5</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>None</td>\n",
       "      <td>[Column name, Column definition, Unnamed: 2, U...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[SampleID, ID of the prospective participant,...</td>\n",
       "      <td>[Description, A-Annotation, B-ClinicalData]</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>729.00</td>\n",
       "      <td>N2b</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>T4a</td>\n",
       "      <td>01CO001</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>Male</td>\n",
       "      <td>[SampleID, Clinical, Mutation, CNV, miRNA-seq,...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...</td>\n",
       "      <td>[Description, A-Annotation, B-ClinicalData]</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Living</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>838.00</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>T3</td>\n",
       "      <td>01CO005</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[SampleID, Clinical, Mutation, CNV, miRNA-seq,...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...</td>\n",
       "      <td>[Description, A-Annotation, B-ClinicalData]</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>904.00</td>\n",
       "      <td>N2b</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>T4a</td>\n",
       "      <td>01CO006</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[SampleID, Clinical, Mutation, CNV, miRNA-seq,...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...</td>\n",
       "      <td>[Description, A-Annotation, B-ClinicalData]</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Ascending Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Living</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>652.00</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>T3</td>\n",
       "      <td>01CO008</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>vasaikar_mmc1.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[SampleID, Clinical, Mutation, CNV, miRNA-seq,...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...</td>\n",
       "      <td>[Description, A-Annotation, B-ClinicalData]</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Descending Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Living</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>None</td>\n",
       "      <td>[README for Table S1, Unnamed: 1, Unnamed: 2]</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[nan, nan, nan], [Tab1, clinical_data, Clinic...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>58.00</td>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>Not available</td>\n",
       "      <td>C3L-00104</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>Male</td>\n",
       "      <td>[case_id, age, gender, height, weight, bmi, co...</td>\n",
       "      <td>...</td>\n",
       "      <td>GBM</td>\n",
       "      <td>Caucasian</td>\n",
       "      <td>[[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "      <td>Wang</td>\n",
       "      <td>Frontal Lobe</td>\n",
       "      <td>unilateral</td>\n",
       "      <td>4</td>\n",
       "      <td>608.0</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>59.00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>Female</td>\n",
       "      <td>[case_id, age, gender, height, weight, bmi, co...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>45.00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>None</td>\n",
       "      <td>[case_id, age, gender, height, weight, bmi, co...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>69.00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>None</td>\n",
       "      <td>[case_id, age, gender, height, weight, bmi, co...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>77.00</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>None</td>\n",
       "      <td>[case_id, age, gender, height, weight, bmi, co...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Living</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C3L-00104</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>None</td>\n",
       "      <td>[case, sample_type, multiomic, nmf_consensus, ...</td>\n",
       "      <td>...</td>\n",
       "      <td>GBM, RTK II</td>\n",
       "      <td>None</td>\n",
       "      <td>[[C3L-00104, tumor, nmf1, nmf1, 0.743, Proneur...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "      <td>Wang</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>2632.0</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...</td>\n",
       "      <td>None</td>\n",
       "      <td>wang_mmc2.xlsx</td>\n",
       "      <td>None</td>\n",
       "      <td>[aliquot_id, sample_id, original_aliquot_id, b...</td>\n",
       "      <td>...</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>[[7316-288-TISS [894482], 7316-288, 7316-288-T...</td>\n",
       "      <td>[README, clinical_data, additional_annotations...</td>\n",
       "      <td>Wang</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Alive</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>18 rows Ã— 23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    age_at_diagnosis ajcc_pathologic_n ajcc_pathologic_stage  \\\n",
       "0              38.88               pN0               Stage I   \n",
       "1              39.76               pNX              Stage IV   \n",
       "2              51.19               pN0               Stage I   \n",
       "3                NaN              None                  None   \n",
       "4              32.69               pNX               Stage I   \n",
       "5                NaN              None                  None   \n",
       "6             729.00               N2b             Stage III   \n",
       "7             838.00                N0              Stage II   \n",
       "8             904.00               N2b             Stage III   \n",
       "9             652.00                N0              Stage II   \n",
       "10               NaN              None                  None   \n",
       "11             58.00                                           \n",
       "12             59.00              None                  None   \n",
       "13             45.00              None                  None   \n",
       "14             69.00              None                  None   \n",
       "15             77.00              None                  None   \n",
       "16               NaN              None                  None   \n",
       "17               NaN              None                  None   \n",
       "\n",
       "   ajcc_pathologic_t case_submitter_id  \\\n",
       "0     pT1a (FIGO IA)         C3L-00006   \n",
       "1     pT1a (FIGO IA)         C3L-00008   \n",
       "2     pT1a (FIGO IA)         C3L-00032   \n",
       "3               None         C3L-00084   \n",
       "4     pT1a (FIGO IA)         C3L-00090   \n",
       "5               None              None   \n",
       "6                T4a           01CO001   \n",
       "7                 T3           01CO005   \n",
       "8                T4a           01CO006   \n",
       "9                 T3           01CO008   \n",
       "10              None              None   \n",
       "11     Not available         C3L-00104   \n",
       "12              None              None   \n",
       "13              None              None   \n",
       "14              None              None   \n",
       "15              None              None   \n",
       "16              None         C3L-00104   \n",
       "17              None              None   \n",
       "\n",
       "                                             contents               ethnicity  \\\n",
       "0   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  Not-Hispanic or Latino   \n",
       "1   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  Not-Hispanic or Latino   \n",
       "2   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  Not-Hispanic or Latino   \n",
       "3   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                      No   \n",
       "4   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  Not-Hispanic or Latino   \n",
       "5   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "6   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "7   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "8   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "9   b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "10  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "11  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...  Not-Hispanic or Latino   \n",
       "12  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "13  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "14  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "15  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "16  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "17  b'PK\\x03\\x04\\x14\\x00\\x06\\x00\\x08\\x00\\x00\\x00!\\...                    None   \n",
       "\n",
       "              filename  gender  \\\n",
       "0        dou_mmc1.xlsx  Female   \n",
       "1        dou_mmc1.xlsx  Female   \n",
       "2        dou_mmc1.xlsx  Female   \n",
       "3        dou_mmc1.xlsx    None   \n",
       "4        dou_mmc1.xlsx  Female   \n",
       "5   vasaikar_mmc1.xlsx    None   \n",
       "6   vasaikar_mmc1.xlsx    Male   \n",
       "7   vasaikar_mmc1.xlsx  Female   \n",
       "8   vasaikar_mmc1.xlsx  Female   \n",
       "9   vasaikar_mmc1.xlsx  Female   \n",
       "10      wang_mmc2.xlsx    None   \n",
       "11      wang_mmc2.xlsx    Male   \n",
       "12      wang_mmc2.xlsx  Female   \n",
       "13      wang_mmc2.xlsx    None   \n",
       "14      wang_mmc2.xlsx    None   \n",
       "15      wang_mmc2.xlsx    None   \n",
       "16      wang_mmc2.xlsx    None   \n",
       "17      wang_mmc2.xlsx    None   \n",
       "\n",
       "                                               header  ... primary_diagnosis  \\\n",
       "0   [idx, Proteomics_Participant_ID, Case_excluded...  ...      Endometrioid   \n",
       "1   [idx, Proteomics_Participant_ID, Case_excluded...  ...      Endometrioid   \n",
       "2   [idx, Proteomics_Participant_ID, Case_excluded...  ...      Endometrioid   \n",
       "3   [idx, Proteomics_Participant_ID, Case_excluded...  ...    Carcinosarcoma   \n",
       "4   [idx, Proteomics_Participant_ID, Case_excluded...  ...      Endometrioid   \n",
       "5   [Column name, Column definition, Unnamed: 2, U...  ...              None   \n",
       "6   [SampleID, Clinical, Mutation, CNV, miRNA-seq,...  ...              None   \n",
       "7   [SampleID, Clinical, Mutation, CNV, miRNA-seq,...  ...              None   \n",
       "8   [SampleID, Clinical, Mutation, CNV, miRNA-seq,...  ...              None   \n",
       "9   [SampleID, Clinical, Mutation, CNV, miRNA-seq,...  ...              None   \n",
       "10      [README for Table S1, Unnamed: 1, Unnamed: 2]  ...              None   \n",
       "11  [case_id, age, gender, height, weight, bmi, co...  ...               GBM   \n",
       "12  [case_id, age, gender, height, weight, bmi, co...  ...              None   \n",
       "13  [case_id, age, gender, height, weight, bmi, co...  ...              None   \n",
       "14  [case_id, age, gender, height, weight, bmi, co...  ...              None   \n",
       "15  [case_id, age, gender, height, weight, bmi, co...  ...              None   \n",
       "16  [case, sample_type, multiomic, nmf_consensus, ...  ...       GBM, RTK II   \n",
       "17  [aliquot_id, sample_id, original_aliquot_id, b...  ...              None   \n",
       "\n",
       "         race                                               rows  \\\n",
       "0       White  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "1       White  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "2       White  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "3        None  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "4       White  [[S001, C3L-00006, No, 2, 5, 128N, C3L-00006-0...   \n",
       "5        None  [[SampleID, ID of the prospective participant,...   \n",
       "6        None  [[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...   \n",
       "7        None  [[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...   \n",
       "8        None  [[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...   \n",
       "9        None  [[Type, BIN, BIN, BIN, BIN, BIN, BIN, BIN, BIN...   \n",
       "10       None  [[nan, nan, nan], [Tab1, clinical_data, Clinic...   \n",
       "11  Caucasian  [[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...   \n",
       "12       None  [[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...   \n",
       "13       None  [[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...   \n",
       "14       None  [[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...   \n",
       "15       None  [[C3L-00104, 58, Male, 188.0, 115.0, 32.54, Un...   \n",
       "16       None  [[C3L-00104, tumor, nmf1, nmf1, 0.743, Proneur...   \n",
       "17       None  [[7316-288-TISS [894482], 7316-288, 7316-288-T...   \n",
       "\n",
       "                                          sheet_names     study  \\\n",
       "0                       [UCEC_CPTAC3_meta_table_V2.1]  dou_mmc1   \n",
       "1                       [UCEC_CPTAC3_meta_table_V2.1]  dou_mmc1   \n",
       "2                       [UCEC_CPTAC3_meta_table_V2.1]  dou_mmc1   \n",
       "3                       [UCEC_CPTAC3_meta_table_V2.1]  dou_mmc1   \n",
       "4                       [UCEC_CPTAC3_meta_table_V2.1]  dou_mmc1   \n",
       "5         [Description, A-Annotation, B-ClinicalData]      None   \n",
       "6         [Description, A-Annotation, B-ClinicalData]  Vasaikar   \n",
       "7         [Description, A-Annotation, B-ClinicalData]  Vasaikar   \n",
       "8         [Description, A-Annotation, B-ClinicalData]  Vasaikar   \n",
       "9         [Description, A-Annotation, B-ClinicalData]  Vasaikar   \n",
       "10  [README, clinical_data, additional_annotations...      None   \n",
       "11  [README, clinical_data, additional_annotations...      Wang   \n",
       "12  [README, clinical_data, additional_annotations...      None   \n",
       "13  [README, clinical_data, additional_annotations...      None   \n",
       "14  [README, clinical_data, additional_annotations...      None   \n",
       "15  [README, clinical_data, additional_annotations...      None   \n",
       "16  [README, clinical_data, additional_annotations...      Wang   \n",
       "17  [README, clinical_data, additional_annotations...      Wang   \n",
       "\n",
       "             tissue_or_organ_of_origin tumor_focality   tumor_grade  \\\n",
       "0                 Anterior endometrium       Unifocal  FIGO grade 1   \n",
       "1                Posterior endometrium       Unifocal  FIGO grade 1   \n",
       "2   Anterior and Posterior endometrium       Unifocal  FIGO grade 2   \n",
       "3                                 None           None          None   \n",
       "4   Anterior and Posterior endometrium       Unifocal  FIGO grade 2   \n",
       "5                                 None           None          None   \n",
       "6                        Sigmoid Colon           None          None   \n",
       "7                        Sigmoid Colon           None          None   \n",
       "8                      Ascending Colon           None          None   \n",
       "9                     Descending Colon           None          None   \n",
       "10                                None           None          None   \n",
       "11                        Frontal Lobe     unilateral             4   \n",
       "12                                None           None          None   \n",
       "13                                None           None          None   \n",
       "14                                None           None          None   \n",
       "15                                None           None          None   \n",
       "16                                None           None          None   \n",
       "17                                None           None          None   \n",
       "\n",
       "   tumor_largest_dimension_diameter vital_status  \n",
       "0                               2.9           No  \n",
       "1                               3.5           No  \n",
       "2                               4.5           No  \n",
       "3                              None          Yes  \n",
       "4                               3.5           No  \n",
       "5                              None         None  \n",
       "6                              None       Living  \n",
       "7                              None     Deceased  \n",
       "8                              None       Living  \n",
       "9                              None       Living  \n",
       "10                             None         None  \n",
       "11                            608.0     Deceased  \n",
       "12                             None     Deceased  \n",
       "13                             None     Deceased  \n",
       "14                             None     Deceased  \n",
       "15                             None       Living  \n",
       "16                           2632.0         None  \n",
       "17                             None        Alive  \n",
       "\n",
       "[18 rows x 23 columns]"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Chosen plan\n",
      " 0. MarshalAndScanDataOp -> XLSFile \n",
      "\n",
      " 1. XLSFile -> NonLLMConvert -> Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']]\n",
      "    (contents, filename, number_she) -> (contents, filename, header, na)\n",
      "    UDF: <function xls_to_tables at 0x7fcdc4214720>\n",
      "\n",
      " 2. Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']] -> LLMFilter -> Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']]\n",
      "    (contents, filename, header, na) -> (contents, filename, header, na)\n",
      "    Model: Model.GPT_4o_MINI\n",
      "    Filter: The table contains biometric information about the patient\n",
      "\n",
      " 3. Schema[['contents', 'filename', 'filename', 'header', 'name', 'number_sheets', 'rows', 'sheet_names']] -> LLMConvertBonded -> Schema[['age_at_diagnosis', 'ajcc_pathologic_n', 'ajcc_pathologic_stage', 'ajcc_pathologic_t', 'case_submitter_id', 'contents', 'ethnicity', 'filename', 'gender', 'header', 'morphology', 'name', 'number_sheets', 'primary_diagnosis', 'race', 'rows', 'sheet_names', 'study', 'tissue_or_organ_of_origin', 'tumor_focality', 'tumor_grade', 'tumor_largest_dimension_diameter', 'vital_status']]\n",
      "    (contents, filename, header, na) -> (age_at_diagnosis, ajcc_patholo)\n",
      "    Model: Model.GPT_4o_MINI\n",
      "    Prompt Strategy: PromptStrategy.COT_QA\n",
      "\n",
      "\n",
      "Total execution time: 100.12398648262024 seconds\n",
      "Total execution cost: 0.014461200000000002 USD\n"
     ]
    }
   ],
   "source": [
    "papers_html = pz.Dataset(\"../testdata/biofabric-html-tiny\")\n",
    "table_urls = papers_html.sem_add_columns(\n",
    "    [{\"name\": \"url\", \"type\": str, \"desc\": \"The URLs of the XLS tables from the page\"}],\n",
    "    cardinality=pz.Cardinality.ONE_TO_MANY)\n",
    "files = table_urls.add_columns(udf=udfs.url_to_file, cols=file_cols)\n",
    "# xls = files.download()\n",
    "xls = pz.Dataset(\"../testdata/biofabric-tiny\")\n",
    "tables = xls.add_columns(udf=udfs.xls_to_tables, cols=table_cols, cardinality=pz.Cardinality.ONE_TO_MANY)\n",
    "patient_tables = tables.sem_filter(\"The table contains biometric information about the patient\")\n",
    "case_data = patient_tables.sem_add_columns(case_data_cols, cardinality=pz.Cardinality.ONE_TO_MANY)\n",
    "\n",
    "policy = pz.MinCost()\n",
    "config = pz.QueryProcessorConfig(policy=policy, nocache=True)\n",
    "result = case_data.run(config)\n",
    "\n",
    "print_result(result.data_records)\n",
    "print_stats(result.execution_stats)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_table.__dict__.keys()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Maximum Cost: GPT 4"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "LOGICAL PLANS: 1\n",
      "INITIAL PLANS: 27\n",
      "DEDUP PLANS: 27\n",
      "PARETO PLANS: 17\n",
      "----------\n",
      "Policy is: Maximum Quality\n",
      "Chosen plan: Time est: 411.647 -- Cost est: 0.973 -- Quality est: 0.486\n",
      " 0. MarshalAndScanDataOp -> File \n",
      "\n",
      " 1. File -> InduceFromCandidateOp -> XLSFile \n",
      "    Using Model.GPT_3_5\n",
      "    (contents,filena...) -> (contents,filena...)\n",
      "\n",
      " 2. XLSFile -> InduceFromCandidateOp -> Table \n",
      "    Using Model.GPT_4\n",
      "    (contents,filena...) -> (filename,header...)\n",
      "\n",
      " 3. Table -> FilterCandidateOp -> Table \n",
      "    Using Model.GPT_4\n",
      "    Filter: \"The table contains biometric information about the patient\"\n",
      "    (filename,header...) -> (filename,header...)\n",
      "\n",
      " 4. Table -> InduceFromCandidateOp -> CaseData \n",
      "    Using Model.GPT_4\n",
      "    (filename,header...) -> (age_at_diagnosi...)\n",
      "\n",
      "Bonded query processing error: No output objects were generated with bonded query - trying with conventional query...\n",
      "BondedQuery Error: No output objects were generated with bonded query - trying with conventional query...\n",
      "Falling back to conventional query\n",
      "Could not find any items in the JSON response\n",
      "Could not find any items in the JSON response\n",
      "Bonded query processing error: No output objects were generated with bonded query - trying with conventional query...\n",
      "BondedQuery Error: No output objects were generated with bonded query - trying with conventional query...\n",
      "Falling back to conventional query\n",
      "Could not find any items in the JSON response\n",
      "Could not find any items in the JSON response\n",
      "Could not find any items in the JSON response\n",
      "Could not find any items in the JSON response\n",
      "Could not find any items in the JSON response\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>age_at_diagnosis</th>\n",
       "      <th>ajcc_pathologic_n</th>\n",
       "      <th>ajcc_pathologic_stage</th>\n",
       "      <th>ajcc_pathologic_t</th>\n",
       "      <th>case_submitter_id</th>\n",
       "      <th>ethnicity</th>\n",
       "      <th>gender</th>\n",
       "      <th>morphology</th>\n",
       "      <th>primary_diagnosis</th>\n",
       "      <th>race</th>\n",
       "      <th>study</th>\n",
       "      <th>tissue_or_organ_of_origin</th>\n",
       "      <th>tumor_focality</th>\n",
       "      <th>tumor_grade</th>\n",
       "      <th>tumor_largest_dimension_diameter</th>\n",
       "      <th>vital_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>64.0</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00006</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>White</td>\n",
       "      <td>UCEC_CPTAC3</td>\n",
       "      <td>Anterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>2.9</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>58.0</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Stage IV</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00008</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>FIGO grade 1</td>\n",
       "      <td>White</td>\n",
       "      <td>UCEC_CPTAC3</td>\n",
       "      <td>Posterior endometrium</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>3.5</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>50.0</td>\n",
       "      <td>pN0</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00032</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>White</td>\n",
       "      <td>UCEC_CPTAC3</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>4.5</td>\n",
       "      <td>Yes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C3L-00084</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Carcinosarcoma</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>UCEC_CPTAC3</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>75.0</td>\n",
       "      <td>pNX</td>\n",
       "      <td>Stage I</td>\n",
       "      <td>pT1a (FIGO IA)</td>\n",
       "      <td>C3L-00090</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>Endometrioid</td>\n",
       "      <td>FIGO grade 2</td>\n",
       "      <td>White</td>\n",
       "      <td>UCEC_CPTAC3</td>\n",
       "      <td>Other, specify</td>\n",
       "      <td>Unifocal</td>\n",
       "      <td>Cannot be determined</td>\n",
       "      <td>3.5</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>729.0</td>\n",
       "      <td>N2b</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>T4a</td>\n",
       "      <td>01CO001</td>\n",
       "      <td>None</td>\n",
       "      <td>Male</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Living</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>838.0</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>T3</td>\n",
       "      <td>01CO005</td>\n",
       "      <td>None</td>\n",
       "      <td>Female</td>\n",
       "      <td>Not Mucinous</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Sigmoid Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>904.0</td>\n",
       "      <td>N2b</td>\n",
       "      <td>Stage III</td>\n",
       "      <td>T4a</td>\n",
       "      <td>01CO006</td>\n",
       "      <td>None</td>\n",
       "      <td>Female</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Ascending Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Living</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>652.0</td>\n",
       "      <td>N0</td>\n",
       "      <td>Stage II</td>\n",
       "      <td>T3</td>\n",
       "      <td>01CO008</td>\n",
       "      <td>None</td>\n",
       "      <td>Female</td>\n",
       "      <td>Mucinous</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>Vasaikar</td>\n",
       "      <td>Descending Colon</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Living</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>58.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C3L-00104</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Male</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>White</td>\n",
       "      <td>wang</td>\n",
       "      <td>Frontal Lobe</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>59.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C3L-00365</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>White</td>\n",
       "      <td>wang</td>\n",
       "      <td>Parietal Lobe</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>45.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C3L-00674</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Male</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>White</td>\n",
       "      <td>wang</td>\n",
       "      <td>Frontal Lobe</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>69.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C3L-00677</td>\n",
       "      <td>Not-Hispanic or Latino</td>\n",
       "      <td>Female</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>White</td>\n",
       "      <td>wang</td>\n",
       "      <td>Frontal Lobe</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Deceased</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>77.0</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>C3L-01040</td>\n",
       "      <td>nan</td>\n",
       "      <td>Male</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>nan</td>\n",
       "      <td>wang</td>\n",
       "      <td>Frontal Lobe</td>\n",
       "      <td>None</td>\n",
       "      <td>None</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Living</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    age_at_diagnosis ajcc_pathologic_n ajcc_pathologic_stage  \\\n",
       "0               64.0               pN0               Stage I   \n",
       "1               58.0               pNX              Stage IV   \n",
       "2               50.0               pN0               Stage I   \n",
       "3                NaN              None                  None   \n",
       "4               75.0               pNX               Stage I   \n",
       "5              729.0               N2b             Stage III   \n",
       "6              838.0                N0              Stage II   \n",
       "7              904.0               N2b             Stage III   \n",
       "8              652.0                N0              Stage II   \n",
       "9               58.0              None                  None   \n",
       "10              59.0              None                  None   \n",
       "11              45.0              None                  None   \n",
       "12              69.0              None                  None   \n",
       "13              77.0              None                  None   \n",
       "\n",
       "   ajcc_pathologic_t case_submitter_id               ethnicity  gender  \\\n",
       "0     pT1a (FIGO IA)         C3L-00006  Not-Hispanic or Latino  Female   \n",
       "1     pT1a (FIGO IA)         C3L-00008  Not-Hispanic or Latino  Female   \n",
       "2     pT1a (FIGO IA)         C3L-00032  Not-Hispanic or Latino  Female   \n",
       "3               None         C3L-00084                    None    None   \n",
       "4     pT1a (FIGO IA)         C3L-00090  Not-Hispanic or Latino  Female   \n",
       "5                T4a           01CO001                    None    Male   \n",
       "6                 T3           01CO005                    None  Female   \n",
       "7                T4a           01CO006                    None  Female   \n",
       "8                 T3           01CO008                    None  Female   \n",
       "9               None         C3L-00104  Not-Hispanic or Latino    Male   \n",
       "10              None         C3L-00365  Not-Hispanic or Latino  Female   \n",
       "11              None         C3L-00674  Not-Hispanic or Latino    Male   \n",
       "12              None         C3L-00677  Not-Hispanic or Latino  Female   \n",
       "13              None         C3L-01040                     nan    Male   \n",
       "\n",
       "        morphology primary_diagnosis   race        study  \\\n",
       "0     Endometrioid      FIGO grade 1  White  UCEC_CPTAC3   \n",
       "1     Endometrioid      FIGO grade 1  White  UCEC_CPTAC3   \n",
       "2     Endometrioid      FIGO grade 2  White  UCEC_CPTAC3   \n",
       "3   Carcinosarcoma              None   None  UCEC_CPTAC3   \n",
       "4     Endometrioid      FIGO grade 2  White  UCEC_CPTAC3   \n",
       "5         Mucinous              None   None     Vasaikar   \n",
       "6     Not Mucinous              None   None     Vasaikar   \n",
       "7         Mucinous              None   None     Vasaikar   \n",
       "8         Mucinous              None   None     Vasaikar   \n",
       "9             None              None  White         wang   \n",
       "10            None              None  White         wang   \n",
       "11            None              None  White         wang   \n",
       "12            None              None  White         wang   \n",
       "13            None              None    nan         wang   \n",
       "\n",
       "   tissue_or_organ_of_origin tumor_focality           tumor_grade  \\\n",
       "0       Anterior endometrium       Unifocal  Cannot be determined   \n",
       "1      Posterior endometrium       Unifocal  Cannot be determined   \n",
       "2             Other, specify       Unifocal  Cannot be determined   \n",
       "3                       None           None                  None   \n",
       "4             Other, specify       Unifocal  Cannot be determined   \n",
       "5              Sigmoid Colon           None                  None   \n",
       "6              Sigmoid Colon           None                  None   \n",
       "7            Ascending Colon           None                  None   \n",
       "8           Descending Colon           None                  None   \n",
       "9               Frontal Lobe           None                  None   \n",
       "10             Parietal Lobe           None                  None   \n",
       "11              Frontal Lobe           None                  None   \n",
       "12              Frontal Lobe           None                  None   \n",
       "13              Frontal Lobe           None                  None   \n",
       "\n",
       "    tumor_largest_dimension_diameter vital_status  \n",
       "0                                2.9           No  \n",
       "1                                3.5           No  \n",
       "2                                4.5          Yes  \n",
       "3                                NaN         None  \n",
       "4                                3.5           No  \n",
       "5                                NaN       Living  \n",
       "6                                NaN     Deceased  \n",
       "7                                NaN       Living  \n",
       "8                                NaN       Living  \n",
       "9                                NaN     Deceased  \n",
       "10                               NaN     Deceased  \n",
       "11                               NaN     Deceased  \n",
       "12                               NaN     Deceased  \n",
       "13                               NaN       Living  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "xls = pz.Dataset(\"testdata/biofabric-tiny\")\n",
    "patient_tables = xls.sem_add_columns(table_cols, cardinality=pz.Cardinality.ONE_TO_MANY)\n",
    "patient_tables = patient_tables.sem_filter(\"The table contains biometric information about the patient\")\n",
    "case_data = patient_tables.sem_add_columns(case_data_cols, cardinality=pz.Cardinality.ONE_TO_MANY)\n",
    "\n",
    "policy = pz.MaxQuality()\n",
    "config = pz.QueryProcessorConfig(\n",
    "    policy=policy,\n",
    "    nocache=True,\n",
    "    allow_token_reduction=False,\n",
    "    allow_code_synth=False,\n",
    "    processing_strategy=\"streaming\",\n",
    "    execution_strategy=\"sequential\",\n",
    ")\n",
    "iterable = case_data.run(config)\n",
    "\n",
    "output_rows = []\n",
    "for data_record_collection in iterable:  # noqa: B007\n",
    "    for output_table in data_record_collection:\n",
    "        print(output_table.to_dict().keys())\n",
    "        output_rows.append(output_table.to_dict()) \n",
    "\n",
    "output_df = pd.DataFrame(output_rows)\n",
    "display(output_df)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "palimpzest",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
